application/controllers/FailedEmailController.php

```diff
@@ -0,0 +1,213 @@
+
+        return parent::beforeRender($view);
+    }
+
+    /**
+     * @throws CHttpException|CException
+     */
+    public function actionIndex(): void
+    {
+        $surveyId = sanitize_int(App()->request->getParam('surveyid'));
+        $permissions = [
+            'update' => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'update'),
+            'delete' => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'delete'),
+            'read'   => Permission::model()->hasSurveyPermission($surveyId, 'responses', 'read')
+        ];
+        if (!$surveyId) {
+            throw new CHttpException(403, gT("Invalid survey ID"));
+        }
+        if (!$permissions['read']) {
+            App()->user->setFlash('error', gT("You do not have permission to access this page."));
+            $this->redirect(['surveyAdministration/view', 'surveyid' => $surveyId]);
+        }
+
+        App()->getClientScript()->registerScriptFile('/application/views/failedEmail/javascript/failedEmail.js', LSYii_ClientScript::POS_BEGIN);
+        $failedEmailModel = FailedEmail::model();
+        $failedEmailModel->setAttributes(App()->getRequest()->getParam('FailedEmail'), false);
+        $failedEmailModel->setAttribute('surveyid', $surveyId);
+        $pageSize = App()->request->getParam('pageSize') ?? App()->user->getState('pageSize', App()->params['defaultPageSize']);
+        $massiveAction = App()->getController()->renderPartial('/failedEmail/partials/massive_action_selector', [
+            'surveyId' => $surveyId,
+            'permissions' => $permissions
+        ], true);
+
+
+        $this->render('failedEmail_index', [
+            'failedEmailModel' => $failedEmailModel,
+            'pageSize'         => $pageSize,
+            'massiveAction'    => $massiveAction,
+        ]);
+    }
+
+    /**
+     * @throws CHttpException|CException
+     * @return string|void
+     */
+    public function actionResend()
+    {
+        $surveyId = (int)sanitize_int(App()->request->getParam('surveyid'));
+        if (!$surveyId) {
+            throw new CHttpException(403, gT("Invalid survey ID"));
+        }
+        if (!Permission::model()->hasSurveyPermission($surveyId, 'responses', 'update')) {
+            App()->user->setFlash('error', gT("You do not have permission to access this page."));
+            $this->redirect(['failedEmail/index/', 'surveyid' => $surveyId]);
+        }
+        $preserveResend = App()->request->getParam('preserveResend') ?? false;
+        $item = [App()->request->getParam('item')];
+        $items = json_decode(App()->request->getParam('sItems'));
+        $selectedItems = $items ?? $item;
+        $emailsByType = [];
+        if (!empty($selectedItems)) {
+            $criteria = new CDbCriteria();
+            $criteria->select = 'id, email_type, recipient';
+            $criteria->addCondition('surveyid = ' . (int) $surveyId);
```

Comment:
`$criteria->compare('surveyid', $surveyId):`

Reply:
Does it matter? If you suggest a change, feel free to add motivation. :)

Reply:
motivation : usage of PDO mean usage of more SQL cache system.

Reply:
Isn't PDO always used by Yii internally?

Reply:
Sorry : PDO with parameters :)

Else : if `(int) $surveyId != $surveyId` it must throw an error , no ?

Reply:
But this one is not important, justthink is better.
Have the 2 new function in Authentication controller are really more strange …

Reply:
Oh yeah, prepared statements. Yes, if `$criteria->compare('surveyid', $surveyId):` is doing that, then it's better.

Reply:
lulz
https://github.com/yiisoft/yii/blob/master/framework/db/schema/CDbCriteria.php#L446